### PR TITLE
add some JMS attributes

### PR DESCRIPTION
**Pull request by malafeev**
I'm trying to add what I can from https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/messaging.md#messaging-attributes

But I don't see the way for some attributes like:
- messaging.system
- messaging.protocol
- messaging.protocol_version
- messaging.url
- messaging.message_payload_size_bytes
- messaging.message_payload_compressed_size_bytes

any ideas?

### CODE REVIEW

File: instrumentation/jms-1.1/src/main/java/io/opentelemetry/auto/instrumentation/jms/JMSDecorator.java [hunk 1]
```diff
@@ -80,4 +83,35 @@ public static String toSpanName(Destination destination) {
...
+    if (spanName.startsWith("queue/")) {
+      SemanticAttributes.MESSAGING_DESTINATION_KIND.set(span, "queue");
+      SemanticAttributes.MESSAGING_DESTINATION.set(span, spanName.replaceFirst("queue/", ""));
+    } else if (spanName.startsWith("topic/")) {
+      SemanticAttributes.MESSAGING_DESTINATION_KIND.set(span, "topic");
+      SemanticAttributes.MESSAGING_DESTINATION.set(span, spanName.replaceFirst("topic/", ""));
+    }
+    if (spanName.equals("queue/<temporary>") || spanName.equals("topic/<temporary>")) {
+      SemanticAttributes.MESSAGING_TEMP_DESTINATION.set(span, true);
+    }
```
Comment (trask) [inline instrumentation/jms-1.1/src/main/java/io/opentelemetry/auto/instrumentation/jms/JMSDecorator.java:97]: when we move to the new `Tracer` design, and do both `JMSDecorator.toSpanName()` and `afterStart()` inside of `startSpan`, we can consolidate the conditionals between these two methods (and remove the `startsWith` logic)

**Comment by anuraaga** [comment]
Is now a good time to also update the name? Or should we do it in another PR? It was updated to include operation name
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/messaging.md#span-name

**Comment by trask** [comment]
@malafeev I merged upstream into this PR in order to resolve the merge conflicts I created via #970

File: instrumentation/jms-1.1/src/main/java/io/opentelemetry/auto/instrumentation/jms/JMSDecorator.java [hunk 2]
```diff
@@ -80,4 +83,35 @@ public static String toSpanName(Destination destination) {
...
+
+    if (message != null) {
+      try {
+        String messageID = message.getJMSMessageID();
+        if (messageID != null) {
+          SemanticAttributes.MESSAGING_MESSAGE_ID.set(span, messageID);
+        }
+      } catch (JMSException e) {
+      }
```
[the remark this answers is not in the document]
Comment (malafeev) [reply instrumentation/jms-1.1/src/main/java/io/opentelemetry/auto/instrumentation/jms/JMSDecorator.java]: done

File: instrumentation/jms-1.1/src/main/java/io/opentelemetry/auto/instrumentation/jms/JMSMessageListenerInstrumentation.java [hunk 3]
```diff
@@ -79,12 +79,13 @@ public JMSMessageListenerInstrumentation() {
...
 
-      Span.Builder spanBuilder =
-          TRACER.spanBuilder(DECORATE.spanNameForConsumer(message)).setSpanKind(CONSUMER);
+      final String spanName = DECORATE.spanNameForConsumer(message);
+      Span.Builder spanBuilder = TRACER.spanBuilder(spanName).setSpanKind(CONSUMER);
       spanBuilder.setParent(extract(message, GETTER));
 
       Span span = spanBuilder.startSpan();
       DECORATE.afterStart(span);
+      DECORATE.afterStart(span, spanName, message);
```
[the remark this answers is not in the document]
Comment (malafeev) [reply instrumentation/jms-1.1/src/main/java/io/opentelemetry/auto/instrumentation/jms/JMSMessageListenerInstrumentation.java]: done

File: instrumentation/jms-1.1/src/main/java/io/opentelemetry/auto/instrumentation/jms/JMSMessageProducerInstrumentation.java [hunk 4]
```diff
@@ -105,12 +105,10 @@ public static SpanWithScope onEnter(
...
 
-      Span span =
-          TRACER
-              .spanBuilder(DECORATE.spanNameForProducer(message, defaultDestination))
-              .setSpanKind(PRODUCER)
-              .startSpan();
+      final String spanName = DECORATE.spanNameForProducer(message, defaultDestination);
+      Span span = TRACER.spanBuilder(spanName).setSpanKind(PRODUCER).startSpan();
       DECORATE.afterStart(span);
+      DECORATE.afterStart(span, spanName, message);
```
[the remark this answers is not in the document]
Comment (malafeev) [reply instrumentation/jms-1.1/src/main/java/io/opentelemetry/auto/instrumentation/jms/JMSMessageProducerInstrumentation.java]: done

File: instrumentation/jms-1.1/src/main/java/io/opentelemetry/auto/instrumentation/jms/JMSMessageProducerInstrumentation.java [hunk 5]
```diff
@@ -146,12 +144,11 @@ public static SpanWithScope onEnter(
...
-      Span span =
-          TRACER
-              .spanBuilder(DECORATE.spanNameForProducer(message, destination))
-              .setSpanKind(PRODUCER)
-              .startSpan();
+      final String spanName = DECORATE.spanNameForProducer(message, destination);
+
+      Span span = TRACER.spanBuilder(spanName).setSpanKind(PRODUCER).startSpan();
       DECORATE.afterStart(span);
+      DECORATE.afterStart(span, spanName, message);
```
[the remark this answers is not in the document]
Comment (malafeev) [reply instrumentation/jms-1.1/src/main/java/io/opentelemetry/auto/instrumentation/jms/JMSMessageProducerInstrumentation.java]: done

File: instrumentation/jms-1.1/src/test/groovy/SpringTemplateJMS1Test.groovy [hunk 6]
```diff
@@ -66,29 +68,31 @@ class SpringTemplateJMS1Test extends AgentTestRunner {
...
       TextMessage msg = template.receive(destination)
       assert msg.text == messageText
+      msgId.set(msg.getJMSMessageID())
 
       template.send(msg.getJMSReplyTo()) {
         session -> template.getMessageConverter().toMessage("responded!", session)
       }
     }
-    def receivedMessage
+    Message receivedMessage
```
[the remark this answers is not in the document]
Comment (malafeev) [reply instrumentation/jms-1.1/src/test/groovy/SpringTemplateJMS1Test.groovy]: done

File: instrumentation/jms-1.1/src/test/groovy/JMS1Test.groovy [hunk 7]
```diff
@@ -263,6 +281,12 @@ class JMS1Test extends AgentTestRunner {
       }
       errored false
       attributes {
+        "${SemanticAttributes.MESSAGING_DESTINATION_KIND.key()}" destinationType
+        "${SemanticAttributes.MESSAGING_DESTINATION.key()}" destinationName
+        "${SemanticAttributes.MESSAGING_MESSAGE_ID.key()}" messageId
```
[the remark this answers is not in the document]
Comment (malafeev) [reply instrumentation/jms-1.1/src/test/groovy/JMS1Test.groovy]: done

File: instrumentation/jms-1.1/src/test/groovy/SpringListenerJMS1Test.groovy [hunk 8]
```diff
@@ -39,11 +40,13 @@ class SpringListenerJMS1Test extends AgentTestRunner {
     expect:
     assertTraces(2) {
       trace(0, 2) {
-        producerSpan(it, 0, "queue/SpringListenerJMS1")
-        consumerSpan(it, 1, "queue/SpringListenerJMS1", true, MessagingMessageListenerAdapter, span(0))
+        producerSpan(it, 0, "queue", "SpringListenerJMS1")
+        // cannot get messageId therefore get it from span attributes
+        consumerSpan(it, 1, "queue", "SpringListenerJMS1", it.span(1).attributes.get(SemanticAttributes.MESSAGING_MESSAGE_ID.key()).stringValue, true, MessagingMessageListenerAdapter, span(0))
```
[the remark this answers is not in the document]
Comment (malafeev) [reply instrumentation/jms-1.1/src/test/groovy/SpringListenerJMS1Test.groovy]: done

File: instrumentation/jms-1.1/src/main/java/io/opentelemetry/auto/instrumentation/jms/JMSDecorator.java (hunk 1 above)
Comment (malafeev) [reply instrumentation/jms-1.1/src/main/java/io/opentelemetry/auto/instrumentation/jms/JMSDecorator.java]: yeah

**Comment by malafeev** [comment]
@anuraaga if we update operation name then we also need to do what @trask said:

> when we move to the new Tracer design, and do both JMSDecorator.toSpanName() and afterStart() inside of startSpan, we can consolidate the conditionals between these two methods (and remove the startsWith logic)

but I don't understand what `move to the new Tracer design` means and when it will happen.

**Comment by trask** [comment]
I created #983 to track span names for JMS (and to review / update span names for Kafka and RabbitMQ as needed also)

**Comment by trask** [comment]
thx @malafeev!

**Pull request merged**